### PR TITLE
Fix #538; Early stopping

### LINKED ISSUE #538: Early stopping
Provide an ability to automatically stop training in case of performance on validation data deteriorates, or ceases to improve (for a given number consecutive epochs or steps).

Discussed in https://github.com/mozilla/DeepSpeech/issues/534 .

### PR DESCRIPTION
**Pull request by andi4191**
@tilmankamp , @kdavis-mozilla : Request you to have a look at this?

I have added few additional FLAG parameters such as (mean_loss_thresh and std_loss_thresh) to take care of the slight fluctuations in the past n_steps (configurable too).

By default, the early_stop flag has been enabled and validation_step is disabled so it will just show a warning message as:
_W Validation flag needs to be enabled for Early Stopping to work
I STARTING Optimization
I Training of Epoch 0 - loss: 363.021423_

**Test Cases:**
DataSet: LDC93S1
**Variations:** Changed the epochs to 50 and 150 to see if the Early Stopping is being triggered.

**With 50 epochs:**
Early stopping isn't triggered as expected from the logs of losses over Validation dataset

**With 150 epochs:**
Early stopping triggered at epoch 62.

**With validation step as 0 and early_stop as True:**
W Validation flag needs to be enabled for Early Stopping to work
I STARTING Optimization
I Training of Epoch 0 - loss: 363.021423

Please review for any comments. Since the TED dataset has only 10 epochs configured. I feel taking decision on past 4 validation loss was appropriate.

### CODE REVIEW

**Comment by kdavis-mozilla** [comment]
Haven't read the code in detail, but one thing that I don't understand is the interaction with checkpoints and early stopping.

Generally I'm wondering how the best, as measured by loss, checkpoints are retained if say checkpointing takes place every 5 min and only 5 of them are kept while early stopping may decide to stop long after these best checkpoints have been removed.

But again maybe I'm missing something as I just took a quick look.

**Comment by tilmankamp** [comment]
@kdavis-mozilla You are right in regards to checkpointing. This is one of the tough parts, as there is no loss history persisted and therefore cannot be referenced after restoring a checkpoint.

File: DeepSpeech.py [hunk 1]
```diff
@@ -1068,6 +1075,7 @@ def _init(self):
         self._epochs_running = []
         self._epochs_done = []
         self._reset_counters()
+	self._loss = []
```
Comment (tilmankamp) [inline DeepSpeech.py]: indenting

File: DeepSpeech.py [hunk 2]
```diff
@@ -1121,6 +1129,7 @@ def start_coordination(self, data_sets, step=0):
             # We only have to train, if we are told so and are not at the target epoch yet
             self._train = FLAGS.train and self._target_epoch > self._epoch
             self._test = FLAGS.test
+            self._dev = FLAGS.dev
```
Comment (tilmankamp) [inline DeepSpeech.py]: Where is `FLAGS.dev` defined? What's its meaning in this context? If validation is to be done? But that's flagged by `FLAGS.validation_step > 0`.

File: DeepSpeech.py [hunk 3]
```diff
@@ -1150,6 +1159,24 @@ def _next_epoch(self):
         # Indicates, if there were 'new' epoch(s) provided
         result = False
 
+        if self._dev:
```
Comment (tilmankamp) [inline DeepSpeech.py]: So `self._dev` is used as a state machine flag. But then it should not get initialized by a user parameter.

File: DeepSpeech.py [hunk 4]
```diff
@@ -1150,6 +1159,24 @@ def _next_epoch(self):
...
+            if FLAGS.early_stop is True and FLAGS.validation_step > 0:
+                #Check for early stopping condition
+                if(len(self._loss) >= FLAGS.earlystop_nsteps):
+	            mean_loss = np.mean(self._loss[-FLAGS.earlystop_nsteps:-2])
```
Comment (tilmankamp) [inline DeepSpeech.py]: indenting

File: DeepSpeech.py [hunk 5]
```diff
@@ -1150,6 +1159,24 @@ def _next_epoch(self):
...
+	            std_loss = np.std(self._loss[-FLAGS.earlystop_nsteps:-2])
```
Comment (tilmankamp) [inline DeepSpeech.py]: indenting

File: DeepSpeech.py [hunk 6]
```diff
@@ -1150,6 +1159,24 @@ def _next_epoch(self):
...
+                    self._loss = self._loss[-FLAGS.earlystop_nsteps:]
+                    log_debug("Current validation loss: %f, std_of_loss(n_step) :%f mean_loss(n_step): %f" % (self._loss[-1], std_loss, mean_loss))
+                    #Making sure slight fluctuations don't bother the early stopping from performing
+                    if self._loss[-1] > np.max(self._loss[:-2]) or (abs(self._loss[-1] - mean_loss) < FLAGS.estop_mean_thresh and std_loss < FLAGS.estop_std_thresh):
+		        #Time to early stop
```
Comment (tilmankamp) [inline DeepSpeech.py]: indenting

File: DeepSpeech.py [hunk 7]
```diff
@@ -1150,6 +1159,24 @@ def _next_epoch(self):
...
         result = False
 
+        if self._dev:
+            if FLAGS.early_stop is True and FLAGS.validation_step > 0:
+                #Check for early stopping condition
+                if(len(self._loss) >= FLAGS.earlystop_nsteps):
+	            mean_loss = np.mean(self._loss[-FLAGS.earlystop_nsteps:-2])
+	            std_loss = np.std(self._loss[-FLAGS.earlystop_nsteps:-2])
+                    self._loss = self._loss[-FLAGS.earlystop_nsteps:]
+                    log_debug("Current validation loss: %f, std_of_loss(n_step) :%f mean_loss(n_step): %f" % (self._loss[-1], std_loss, mean_loss))
```
Comment (tilmankamp) [inline DeepSpeech.py]: single quotes please :)

File: DeepSpeech.py [hunk 8]
```diff
@@ -1150,6 +1159,24 @@ def _next_epoch(self):
...
+                    #Making sure slight fluctuations don't bother the early stopping from performing
+                    if self._loss[-1] > np.max(self._loss[:-2]) or (abs(self._loss[-1] - mean_loss) < FLAGS.estop_mean_thresh and std_loss < FLAGS.estop_std_thresh):
+		        #Time to early stop
+                        log_info("EarlyStopp triggered!!")
```
Comment (tilmankamp) [inline DeepSpeech.py]: single quotes please :)

File: DeepSpeech.py [hunk 9]
```diff
@@ -1522,6 +1552,10 @@ def end(self, session):
                         job.mean_edit_distance = total_mean_edit_distance / job.steps
                         job.wer, job.samples = calculate_report(report_results)
 
+                    # if the job was for validation dataset then append it to the COORD's _loss for early stop verification
+		    if job.set_name == "dev":
```
Comment (tilmankamp) [inline DeepSpeech.py]: single quotes please :)

File: DeepSpeech.py (hunk 2 above)
Comment (andi4191) [inline DeepSpeech.py]: I intended to define it from User parameter, but I guess it will be better to enable self._dev at the time when I am adding task for Validation part in an epoch_running

"if FLAGS.validation_step > 0 and (FLAGS.validation_step == 1 or self._epoch > 0) and self._epoch % FLAGS.validation_step == 0:
                     # The current epoch should also have a validation part
                     self._epochs_running.append(Epoch(self._epoch, self._num_jobs_dev, set_name='dev', report=is_display_step))
                    # Something like this and then disabling it after I am done with validation part and enabling the self._train again.
                     self._dev = True
                     self._train = False 
"

Would that be a good idea because it would mean playing and toggling with the user defined parameter self._train at the same time for moving in between state machines.

File: DeepSpeech.py (hunk 9 above)
Comment (andi4191) [inline DeepSpeech.py]: will change it :)

File: DeepSpeech.py (hunk 8 above)
Comment (andi4191) [inline DeepSpeech.py]: will change it :)

File: DeepSpeech.py (hunk 6 above)
Comment (andi4191) [inline DeepSpeech.py]: will change it :)

File: DeepSpeech.py (hunk 7 above)
Comment (andi4191) [inline DeepSpeech.py]: will change it :)

File: DeepSpeech.py (hunk 5 above)
Comment (andi4191) [inline DeepSpeech.py]: will change it :)

File: DeepSpeech.py (hunk 4 above)
Comment (andi4191) [inline DeepSpeech.py]: will change it :)

File: DeepSpeech.py (hunk 3 above)
Comment (andi4191) [inline DeepSpeech.py]: Referenced in the above comment

File: DeepSpeech.py (hunk 1 above)
Comment (andi4191) [inline DeepSpeech.py]: will change it :)

File: DeepSpeech.py (hunk 2 above)
Comment (tilmankamp) [inline DeepSpeech.py]: From the old code it is already clear, when a validation runs - you don't have to deal with that at all. Things should be fine. The only thing you have to care about should be the stopping.

File: DeepSpeech.py [hunk 10]
```diff
@@ -1169,6 +1213,10 @@ def _next_epoch(self):
                     # The current epoch should also have a validation part
                     self._epochs_running.append(Epoch(self._epoch, self._num_jobs_dev, set_name='dev', report=is_display_step))
 
+                    # State Machine flag for decision making regarding early stopping in next epoch function call
+                    self._dev = True
```
Comment (tilmankamp) [inline DeepSpeech.py]: Why in the next epoch? What if the next epoch is not a validating epoch? I'd place all early stopping code just here. Then there is also no need for the `self._dev`at all. The only state you'd need is `self._loss`.

File: DeepSpeech.py [hunk 11]
```diff
@@ -1414,6 +1462,9 @@ def train(server=None):
     apply_gradient_op = optimizer.apply_gradients(avg_tower_gradients, global_step=global_step)
 
 
+    if FLAGS.early_stop is True and not FLAGS.validation_step > 0:
+        log_warn('Validation flag needs to be enabled for Early Stopping to work')
```
Comment (tilmankamp) [inline DeepSpeech.py]: I'd mention the command line parameters here.

File: DeepSpeech.py [hunk 12]
```diff
@@ -1068,6 +1081,10 @@ def _init(self):
         self._epochs_running = []
         self._epochs_done = []
         self._reset_counters()
+        self._loss = []
```
Comment (tilmankamp) [inline DeepSpeech.py]: I'd give it a more descriptive name like `_dev_losses` or so.

File: DeepSpeech.py [hunk 13]
```diff
@@ -1150,6 +1167,33 @@ def _next_epoch(self):
...
+                    self._loss = self._loss[-FLAGS.earlystop_nsteps:]
+                    log_debug('Current validation loss: %f, std_of_loss(n_step) :%f mean_loss(n_step): %f' % (self._loss[-1], std_loss, mean_loss))
+
+                    #Making sure slight fluctuations don't bother the early stopping from performing
+                    if self._loss[-1] > np.max(self._loss[:-2]) or (abs(self._loss[-1] - mean_loss) < FLAGS.estop_mean_thresh and std_loss < FLAGS.estop_std_thresh):
+                        #Time to early stop
+                        log_info('EarlyStopp triggered!!')
+            else:
+                #Need to reset the loss from past nsteps
+                self._loss = []
```
Comment (tilmankamp) [inline DeepSpeech.py]: Why?

Comment (andi4191) [inline DeepSpeech.py]: If we go for just checking the previous epoch validation loss and deciding based on that would stop the training even for a slight fluctuation. In this case we can monitor past configured nsteps and see if it deviates slightly or not. Based on that decision is being made whether to stop or not. Thats why the std is taken into account.

File: DeepSpeech.py (hunk 10 above)
Comment (andi4191) [inline DeepSpeech.py]: Its a good point actually, i tried implementing it here but even after the early stopping was triggered there were few train epochs which were being reported. We are accumulating the loss when the case is dev anyway. So it won't report loss on incorrect dataset.

File: DeepSpeech.py (hunk 11 above)
Comment (andi4191) [inline DeepSpeech.py]: Is log message being referred to here?

File: DeepSpeech.py (hunk 12 above)
Comment (andi4191) [inline DeepSpeech.py]: Will change it.

File: DeepSpeech.py (hunk 10 above)
Comment (tilmankamp) [inline DeepSpeech.py]: Ah - I see what you mean. But why do we need a "dev" state at all? Why not checking, if early stopping is activated and enough losses got accumulated in the array and then just check - whatever state the state-machine is in.

File: DeepSpeech.py (hunk 11 above)
Comment (tilmankamp) [inline DeepSpeech.py]: How about
`log_warn('Parameter --validation_step needs to be >0 for early stopping to work')`?

File: DeepSpeech.py (hunk 13 above)
Comment (tilmankamp) [inline DeepSpeech.py]: But why cleaning the array?

File: DeepSpeech.py (hunk 10 above)
Comment (andi4191) [inline DeepSpeech.py]: I thought it would be better to align with the state mahine design rather than checking ever epoch. Hence, enabling the flag only when a validation epoch is enqueued

File: DeepSpeech.py (hunk 13 above)
Comment (andi4191) [inline DeepSpeech.py]: If early stop is not enabled there won't be any point in accumulating the losses same is the case with validation step disabled.

File: DeepSpeech.py (hunk 10 above)
Comment (tilmankamp) [inline DeepSpeech.py]: I'd prefer a simpler and mor robust design.

File: DeepSpeech.py (hunk 11 above)
Comment (andi4191) [inline DeepSpeech.py]: Will change it

File: DeepSpeech.py (hunk 13 above)
Comment (tilmankamp) [inline DeepSpeech.py]: I'd just do nothing then...

File: DeepSpeech.py (hunk 10 above)
Comment (andi4191) [inline DeepSpeech.py]: Ok will change it

File: DeepSpeech.py (hunk 13 above)
Comment (andi4191) [inline DeepSpeech.py]: Ok :)

File: DeepSpeech.py [hunk 14]
```diff
@@ -1150,6 +1164,28 @@ def _next_epoch(self):
         # Indicates, if there were 'new' epoch(s) provided
         result = False
 
+        # Make sure that early stop is enabled and validation part is enabled
+        if FLAGS.early_stop is True and FLAGS.validation_step > 0:
+
+            # Check for early stopping condition
+            if(len(self._dev_losses) >= FLAGS.earlystop_nsteps):
```
Comment (tilmankamp) [inline DeepSpeech.py]: You can combine this `if`with the one before. Saves you one indentation level.

File: DeepSpeech.py [hunk 15]
```diff
@@ -1150,6 +1164,28 @@ def _next_epoch(self):
...
+
+                # Calculate the mean of losses for past epochs
+                mean_loss = np.mean(self._dev_losses[-FLAGS.earlystop_nsteps:-2])
+                # Calculate the standard deviation for losses from validation part in the past epochs
+                std_loss = np.std(self._dev_losses[-FLAGS.earlystop_nsteps:-2])
```
Comment (tilmankamp) [inline DeepSpeech.py]: What's this `-2` for? Why not `-1`?

File: DeepSpeech.py [hunk 16]
```diff
@@ -1150,6 +1164,28 @@ def _next_epoch(self):
...
+                # Update the list of losses incurred
+                self._dev_losses = self._dev_losses[-FLAGS.earlystop_nsteps:]
+                log_debug('Current validation loss: %f, std_of_loss(n_step) :%f mean_loss(n_step): %f' % (self._dev_losses[-1], std_loss, mean_loss))
+
+                # Making sure slight fluctuations don't bother the early stopping from performing
+                if self._dev_losses[-1] > np.max(self._dev_losses[:-2]) or (abs(self._dev_losses[-1] - mean_loss) < FLAGS.estop_mean_thresh and std_loss < FLAGS.estop_std_thresh):
+                    # Time to early stop
+                    log_info('EarlyStop triggered!!')
```
Comment (tilmankamp) [inline DeepSpeech.py]: How about "Early stop triggered."? If at hand: Some additional information (like numbers)?

File: DeepSpeech.py (hunk 15 above)
Comment (andi4191) [inline DeepSpeech.py]: its for picking up the latest error and comparing it with previous n_step losses.

File: DeepSpeech.py (hunk 14 above)
Comment (andi4191) [inline DeepSpeech.py]: It actually crossed my mind earlier. Forgot to change it

File: DeepSpeech.py (hunk 16 above)
Comment (andi4191) [inline DeepSpeech.py]: Numbers as in?

Comment (tilmankamp) [inline DeepSpeech.py]: Yeah - something. Just to reason why it triggered.

Comment (andi4191) [inline DeepSpeech.py]: ok
